subsys/bluetooth/host/sdp.c

```diff
@@ -2302,89 +2302,125 @@ static ssize_t sdp_get_seq_len_item(uint8_t **data, size_t len)
 	return -EMSGSIZE;
 }
 
+static int sdp_loop_seqs(uint8_t **data, size_t len)
+{
+	ssize_t slen;
+	ssize_t pre_slen;
+	uint8_t *end;
+
+	if (len <= 0) {
+		return -EMSGSIZE;
+	}
+
+	pre_slen = -EINVAL;
+	slen = -EINVAL;
+	end = *data + len;
+	/* loop all the SEQ */
+	while (*data < end) {
+		/* how long is current UUID's item data associated to */
+		slen = sdp_get_seq_len_item(data, end - *data);
+		if (slen < 0) {
+			break;
+		}
+		pre_slen = slen;
+	}
+
+	/* return the last seq len */
+	if (pre_slen < 0) {
+		return slen;
+	}
+	return pre_slen;
+}
+
 static int sdp_get_uuid_data(const struct bt_sdp_attr_item *attr,
 			     struct bt_sdp_uuid_desc *pd,
-			     uint16_t proto_profile)
+			     uint16_t proto_profile,
+			     uint16_t proto_profile_index)
 {
 	/* get start address of attribute value */
 	uint8_t *p = attr->val;
 	ssize_t slen;
 
 	BT_ASSERT(p);
 
-	/* Attribute value is a SEQ, get length of parent SEQ frame */
-	slen = sdp_get_seq_len_item(&p, attr->len);
-	if (slen < 0) {
-		return slen;
-	}
-
-	/* start reading stacked UUIDs in analyzed sequences tree */
-	while (p - attr->val < attr->len) {
-		size_t to_end, left = 0;
+	while (proto_profile_index > 0) {
+		/* start reading stacked UUIDs in analyzed sequences tree */
+		while (p - attr->val < attr->len) {
+			size_t to_end, left = 0;
 
-		/* to_end tells how far to the end of input buffer */
-		to_end = attr->len - (p - attr->val);
-		/* how long is current UUID's item data associated to */
-		slen = sdp_get_seq_len_item(&p, to_end);
-		if (slen < 0) {
-			return slen;
-		}
+			/* to_end tells how far to the end of input buffer */
+			to_end = attr->len - (p - attr->val);
+			/* loop all the SEQ, get the last SEQ len */
+			slen = sdp_loop_seqs(&p, to_end);
 
-		/* left tells how far is to the end of current UUID */
-		left = slen;
+			if (slen < 0) {
+				return slen;
+			}
 
-		/* check if at least DTD + UUID16 can be read safely */
-		if (left < 3) {
-			return -EMSGSIZE;
-		}
+			/* left tells how far is to the end of current UUID */
+			left = slen;
 
-		/* check DTD and get stacked UUID value */
-		switch (p[0]) {
-		case BT_SDP_UUID16:
-			memcpy(&pd->uuid16,
-			       BT_UUID_DECLARE_16(sys_get_be16(++p)),
-			       sizeof(struct bt_uuid_16));
-			p += sizeof(uint16_t);
-			left -= sizeof(uint16_t);
-			break;
-		case BT_SDP_UUID32:
-			/* check if valid UUID32 can be read safely */
-			if (left < 5) {
+			/* check if at least DTD + UUID16 can be read safely */
+			if (left < 3) {
 				return -EMSGSIZE;
 			}
 
-			memcpy(&pd->uuid32,
-			       BT_UUID_DECLARE_32(sys_get_be32(++p)),
-			       sizeof(struct bt_uuid_32));
-			p += sizeof(uint32_t);
-			left -= sizeof(uint32_t);
-			break;
-		default:
-			BT_ERR("Invalid/unhandled DTD 0x%02x\n", p[0]);
-			return -EINVAL;
-		}
+			/* check DTD and get stacked UUID value */
+			switch (p[0]) {
```

Comment:
Consider storing `p[0]` in a variable for improved readability:
```
uint8_t dtd = p[0];
```

Reply:
Updated, thanks.